Sources/Remote Logging/Crash Logging/SentryEventJSException.swift

```diff
@@ -0,0 +1,96 @@
+import Foundation
+import Sentry
+
+public struct JSException {
```

Comment:
I wonder if it'd be nicer to change `JSException` and `StacktraceLine` to protocols. That means we can avoid the duplication between this struct implementations here and the `GutenbergJSException` in [gutenberg](https://github.com/WordPress/gutenberg/pull/59221/files#diff-879e47f147cb3b7bcda3382ba5bf1888e7208274b4dddade8582e131314603f0), by adding protocol conformations to `GutenbergJSException` at the app level.

Reply:
That's interesting. The only problem I see is that Gutenberg and Tracks are separate libraries that are disconnected from each other. In this scenario, where should the protocols for `JSException` and `StacktraceLine` be defined?

Reply:
My assumption in the current implementation is that WordPress-iOS needs to act as a proxy to pass the exception between libraries and hence, be in charge of converting the exception from `Gutenberg.JSException` to `Tracks.JSException`. Not sure how to conform a protocol in this case.

Reply:
> where should the protocols for JSException and StacktraceLine be defined?

They should be defined in Tracks. But with protocol, you can remove the initializer and definite all properties as read-only. For example.

```swift
public protocol JSException {
  var type: String { get }
  var message: String { get }
  // ...
}
```

And in the app code, only you need to do is declare protocol conformation:
```swift
import Gutenberg

extension GutenbergJSException: Tracks.JSException {
  // You should not need to write any code here, as long as the property declarations
  // in `Tracks.JSException` protocol and `GutenbergJSException` match.
}

extension GutenbergJSException.StacktraceLine: Tracks.JSException.StacktraceLine {}
```

Reply:
If that works, the app just passes the same object from Gutenberg to Tracks, instead of copying one struct as another.

Reply:
Ah, I see. So, if I understand correctly, WP-iOS would expand `GutenbergJSException` struct defined in the Gutenberg Mobile library to conform to the protocol defined in Tracks 👍. I'll try this, thanks @crazytonyli for the tip 🙇 !

Reply:
@crazytonyli I applied your suggestion in https://github.com/Automattic/Automattic-Tracks-iOS/pull/278/commits/270a150509e28eeb8e8142a0d4304ff7101a9e17 and https://github.com/wordpress-mobile/WordPress-iOS/pull/22655/commits/a1a692f09379dc1e081781d78ae488238eb956a6. The only caveat is that I had to redefine the `stacktrace` property because I encountered issues to match the protocol.

Reply:
Thanks for giving it a go! The issue you ran into may have something to do with Swift's type system, where it requires an exact match on the `stacktrace` property type between protocol and the struct. You should be able to "workaround" it by declaring `stacktrace` property as a generic type.

Here is an example:

```swift
// In Tracks

enum Tracks {
    protocol StacktracLine {}

    protocol JSException {
        // Key change: Add an `associatedtype` to declare the protocol type as a generic type.
        associatedtype SL = StacktracLine
        var stacktrace: [SL] { get }
    }
}

// In gutenberg

struct GutenbergeException {
    struct StacktraceLine {
    }
    var stacktrace: [StacktraceLine]
}

// In WordPress

extension GutenbergeException.StacktraceLine: Tracks.StacktracLine {}
extension GutenbergeException: Tracks.JSException {}

```

Reply:
That's interesting. I saw a similar workaround when reading about protocols and cases like this one. I confirmed it works in  https://github.com/Automattic/Automattic-Tracks-iOS/pull/278/commits/d3694e8a352f8cac3f3c26f73125614408c01abc 🎊, thanks @crazytonyli for the suggestion 🙇 !